Tests/SourceKitLSPTests/LocalSwiftTests.swift

```diff
@@ -1476,4 +1476,49 @@ final class LocalSwiftTests: XCTestCase {
     data = EditorPlaceholder(text)
     XCTAssertNil(data)
   }
+  
+  func testIncrementalParse() throws {
+    let url = URL(fileURLWithPath: "/\(UUID())/a.swift")
+    let uri = DocumentURI(url)
+
+    sk.allowUnexpectedNotification = false
+    
+    sk.send(DidOpenTextDocumentNotification(textDocument: TextDocumentItem(
+      uri: uri,
+      language: .swift,
+      version: 0,
+      text: """
+      func foo() {
+      }
+      class bar {
+      }
+      """
+    )))
+    
+    sk.sendNoteSync(DidChangeTextDocumentNotification(textDocument: .init(uri, version: 1), contentChanges: [
+      .init(range: Range(Position(line: 2, utf16index: 7)), text: "a")
+    ]), { (note: LanguageServerProtocol.Notification<PublishDiagnosticsNotification>) -> Void in
+      log("Received diagnostics for text edit - syntactic")
+    }, { (note: LanguageServerProtocol.Notification<PublishDiagnosticsNotification>) -> Void in
+      log("Received diagnostics for text edit - semantic")
+    })
+    
+    
+    let swiftLanguageServer = connection.server!._languageService(for: uri, .swift, in: connection.server!.workspaceForDocumentOnQueue(uri: uri)!) as! SwiftLanguageServer
+    
+    guard let nodes = swiftLanguageServer.documentReusedNodeCollector[uri]?.nodes else {
+      XCTFail("Can not find reused nodes in \(uri)")
+      return
+    }
```

Comment:
This seems racy to me. IIRC the `SwiftLanguageServer` executes `updateSyntaxTree` asynchronously when receiving the note and thus might not have performed the re-parse when you are getting the re-used nodes. This should be easy to fix if you wait for the first `PublishDiagnosticsNotification`.